channels/chan_websocket.c

```diff
@@ -0,0 +1,1509 @@
+#include "asterisk/lock.h"
+#include "asterisk/mod_format.h"
+#include "asterisk/module.h"
+#include "asterisk/pbx.h"
+#include "asterisk/uuid.h"
+#include "asterisk/timing.h"
+#include "asterisk/translate.h"
+#include "asterisk/websocket_client.h"
+
+static struct ast_websocket_server *ast_ws_server;
+
+static struct ao2_container *instances = NULL;
+
+struct websocket_pvt {
+	enum ast_websocket_type type;
+	struct ast_websocket_client *client;
+	struct ast_websocket *websocket;
+	struct ast_format *native_format;
+	struct ast_codec *native_codec;
+	struct ast_format *slin_format;
+	struct ast_codec *slin_codec;
+	struct ast_channel *channel;
+	struct ast_timer *timer;
+	struct ast_frame silence;
+	struct ast_trans_pvt *translator;
+	AST_LIST_HEAD(, ast_frame) frame_queue;
+	pthread_t outbound_read_thread;
+	size_t bytes_read;
+	size_t leftover_len;
+	char *leftover_data;
+	int no_auto_answer;
+	int optimal_frame_size;
+	int bulk_media_in_progress;
+	int report_queue_drained;
+	int frame_queue_length;
+	int queue_full;
+	int queue_paused;
+	char connection_id[0];
+};
+
+#define MEDIA_WEBSOCKET_OPTIMAL_FRAME_SIZE "MEDIA_WEBSOCKET_OPTIMAL_FRAME_SIZE"
+#define MEDIA_WEBSOCKET_CONNECTION_ID "MEDIA_WEBSOCKET_CONNECTION_ID"
+#define INCOMING_CONNECTION_ID "INCOMING"
+
+#define ANSWER_CHANNEL "ANSWER"
+#define HANGUP_CHANNEL "HANGUP"
+#define START_MEDIA_BUFFERING "START_MEDIA_BUFFERING"
+#define STOP_MEDIA_BUFFERING "STOP_MEDIA_BUFFERING"
+#define FLUSH_MEDIA "FLUSH_MEDIA"
+#define GET_DRIVER_STATUS "GET_STATUS"
+#define REPORT_QUEUE_DRAINED "REPORT_QUEUE_DRAINED"
+#define PAUSE_MEDIA "PAUSE_MEDIA"
+#define CONTINUE_MEDIA "CONTINUE_MEDIA"
+
+#define MEDIA_START "MEDIA_START"
+#define MEDIA_XON "MEDIA_XON"
+#define MEDIA_XOFF "MEDIA_XOFF"
+#define QUEUE_DRAINED "QUEUE_DRAINED"
+#define DRIVER_STATUS "STATUS"
+#define MEDIA_BUFFERING_COMPLETED "MEDIA_BUFFERING_COMPLETED"
+
+#define QUEUE_LENGTH_MAX 1000
+#define QUEUE_LENGTH_XOFF_LEVEL 900
+#define QUEUE_LENGTH_XON_LEVEL 800
+#define MAX_TEXT_MESSAGE_LEN MIN(128, (AST_WEBSOCKET_MAX_RX_PAYLOAD_SIZE - 1))
+
+/* Forward declarations */
+static struct ast_channel *webchan_request(const char *type, struct ast_format_cap *cap, const struct ast_assigned_ids *assignedids, const struct ast_channel *requestor, const char *data, int *cause);
+static int webchan_call(struct ast_channel *ast, const char *dest, int timeout);
+static struct ast_frame *webchan_read(struct ast_channel *ast);
+static int webchan_write(struct ast_channel *ast, struct ast_frame *f);
+static int webchan_hangup(struct ast_channel *ast);
+
+static struct ast_channel_tech websocket_tech = {
+	.type = "WebSocket",
+	.description = "Media over WebSocket Channel Driver",
+	.requester = webchan_request,
+	.call = webchan_call,
+	.read = webchan_read,
+	.write = webchan_write,
+	.hangup = webchan_hangup,
+};
+
+static void set_channel_format(struct websocket_pvt * instance,
+	struct ast_format *fmt)
+{
+	if (!ast_strings_equal(
```

Comment:
Why ast_strings_equal vs ast_format_cmp?

Reply:
Ah, that was a holdover from way back when I was using ast_format_def/ast_filestream to do the frame creation.  Will fix.